Render Pass Optimization from TheForge

darksylinc opened the pull request:
The work was performed by collaboration of TheForge and Google. I am merely splitting it up into smaller PRs and cleaning it up.

Before this PR Godot was doing the following (when rendering to the SHADOW MAP):

To the whole atlas:

1. BeingPass - Clear
2. EndPass - Store (that's it, it just clears the shadow map)
3. Light [i] BeginPass - Load
4. SetViewport( atlas_region[i] )
5. Draw calls.
6. Light [i] EndPass - Store
7. Go to step 3 until all lights have been rendered.

After this PR Godot now does the following:

1. BeginPass - Clear
2. SetViewport( atlas_region[i] )
3. Draw calls.
4. Go to step 2 until all lights have been rendered.
5. EndPass - Store

That's a massive optimization for mobile.

I may even argue we should port this optimization to Forward+ (the impact should be much lower though since Desktops are immediate mode renderers, except macOS & iPhone which are powerful enough to run the Forward+ but are tilers).

However there's also to consider that on Desktop it might be a loss because the Clear gets reordered and internally AMD a Clear means a Compute barrier; thus may change how effective asynchronous async compute ends up being (in practice). But if that's a loss or win, I've got no idea.

I've added a comment to the Forward+ implementation to document why the clear is happening there.

Additionally, Godot's Render Graph now tracks if the draw or compute lists are empty. If there's nothing to be done, it skips work.

### Toggle

 - `bool render_pass_opts_enabled = true`

This boolean reverts the shadow map behavior.

Useful for debugging bugs and regressions.

There's no other reason for this boolean. Possibly once it becomes well tested, the boolean could be removed entirely.

# Video

## Before

https://github.com/user-attachments/assets/e42dc3d1-0c91-41a8-aeed-0f26081a6991


## After

https://github.com/user-attachments/assets/4199600d-d144-4a19-b421-01d6f0981ed4

darksylinc committed:
Render Pass Optimization from TheForge

The work was performed by collaboration of TheForge and Google. I am
merely splitting it up into smaller PRs and cleaning it up.

Before this PR Godot was doing the following (when rendering to the
SHADOW MAP):

To the whole atlas:

1. BeingPass - Clear
2. EndPass - Store (that's it, it just clears the shadow map)
3. Light [i] BeginPass - Load
4. SetViewport( atlas_region[i] )
5. Draw calls.
6. Light [i] EndPass - Store
7. Go to step 3 until all lights have been rendered.

After this PR Godot now does the following:

1. BeginPass - Clear
2. SetViewport( atlas_region[i] )
3. Draw calls.
4. Go to step 2 until all lights have been rendered.
5. EndPass - Store

That's a massive optimization for mobile.

I may even argue we should port this optimization to Forward+ (the
impact should be much lower though since Desktops are immediate mode
renderers, except macOS & iPhone which are powerful enough to run the
Forward+ but are tilers).

However there's also to consider that on Desktop it might be a loss
because the Clear gets reordered and internally AMD a Clear means a
Compute barrier; thus may change how effective asynchronous async
compute ends up being (in practice). But if that's a loss or win, I've
got no idea.

I've added a comment to the Forward+ implementation to document why the
clear is happening there.

Additionally, Godot's Render Graph now tracks if the draw or compute
lists are empty. If there's nothing to be done, it skips work.

### Toggle

 - `bool render_pass_opts_enabled = true`

This boolean reverts the shadow map behavior.

Useful for debugging bugs and regressions.

There's no other reason for this boolean. Possibly once it becomes well
tested, the boolean could be removed entirely.

clayjohn commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -686,10 +686,12 @@ void RenderForwardMobile::_pre_opaque_render(RenderDataRD *p_render_data) {
 			light_storage->update_directional_shadow_atlas();
-			RD::get_singleton()->draw_list_begin(light_storage->direction_shadow_get_fb(), RD::DRAW_CLEAR_DEPTH, Vector<Color>(), 0.0f);
-			RD::get_singleton()->draw_list_end();
+			if (!RenderingDeviceCommons::render_pass_opts_enabled) {
+				// Clear the shadow map now to overlap with some async compute.
+				RD::get_singleton()->draw_list_begin(light_storage->direction_shadow_get_fb(), RD::DRAW_CLEAR_DEPTH, Vector<Color>(), 0.0f);
+				RD::get_singleton()->draw_list_end();
+			}
```
I would just remove this entirely. I see the point of having this to help with performance comparisons. But I would remove it before merging.

darksylinc replied:
This block is mostly to debug regression, because it allows us to toggle `render_pass_opts_enabled = false` and see if that immediately fixes the problem.

Ultimately the goal is to remove the render_pass_opts_enabled boolean once we are sure it's stable.

clayjohn commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -1473,9 +1476,70 @@ void RenderForwardMobile::_render_shadow_process() {
+				}
+				// A cubemap is made of 6 faces
+				case RS::LIGHT_OMNI:
+					pass_amount = 6;
```
Basically, Spotlights + Dual Parabaloid omni lights need the same treatment as DirectionalLights.

The overall flow of shadow rendering is like this
```
- For each face of each omnilight
    - `_render_shadow_append()`
    - `if face % 6 ==5`
         - `_render_shadow_end()`
         - Copy to positional shadow atlas
- Clear the Directional Shadow map
- For each directional light
    - for each split
        - `_render_shadow_append()`
- For each spotlight and each DP omnilight face
    - `_render_shadow_append()`
- `_render_shadow_end()` // Renders all Directional Lights directly to directional atlas, renders all spotlights and DPs to positional atlas
```
So Cube Omnilights will need to have one draw list for each face, but spotlights and DPs should be able to all share a draw list after Directional Lights have finished. 

Currently, it works this awkward way because each light is responsible for clearing its own section. So maybe we should change the order and do Spotlights and DPs first and cubes after. Then we can open a draw list with a full clear, render all spots and DPs. Close the draw list. Then render all cube omnilights and copy them over. What do you think?

darksylinc replied:
OK, I've reviewed these remarks.

Well spotted! It was "working fine" because the loop later does:

```
for (uint32_t j = 0; j < pass_amount && curr_shadow_index + j < scene_state.shadow_passes.size(); j++) {
}
```

And the `curr_shadow_index + j < scene_state.shadow_passes.size()` part was hiding the bug from me during execution, as it was early-outing the loop before pass_amount >= 2. But this would've certainly break on other projects.

I just fixed this bug.

> Currently, it works this awkward way because each light is responsible for clearing its own section. So maybe we should change the order and do Spotlights and DPs first and cubes after.

This is problematic because quadrants may not be updated if nothing changed. For what you want to do to work, we would need to do the following:

Let's say we have 4 spotlights, 512x512 each:

1. Clear 1024x1024 region
2. SetViewport to the first 512x512 block
3. Render
4. Go to step for the next 512x512 until all 4 blocks are rendered
5. Close draw list

The problem with that, is that if a single spotlight in those 4 blocks doesn't want to be updated, we must still update it or else the 1024x1024 clear will erase the 512x512 region.

So either we're extremely smart in how we group lights (e.g. clear & render to a 1024x512 region, then clear & render to a 512x512 region), or we force lights to update even if they don't have to if they're caught in the middle (or copy those lights to a temporary location so we can then restore their contents).

It would complicate the logic behind it, a lot.

clayjohn replied:
Oooooh, right. I forgot how aggressively we cache shadows. Indeed, ideally we only update a small subset of lights per frame, so we shouldn't clear the entire atlas.

darksylinc commented:
Ugh. I'll check out what's going on.

40 -> 36 FPS is not a minor regression. That's a huge 2.77ms difference.

It's probably some overlapping thing. Even though Mobile doesn't have async compute, they still try to overlap VS and PS workloads.

I assume this was TPS in Mobile renderer right? (I know it's a stupid question, but I've gotta ask).

clayjohn commented:
> I assume this was TPS in Mobile renderer right? (I know it's a stupid question, but I've gotta ask).

Yep, TPS demo, mobile renderer. Using the version from the Forge PR, but with a Directional Light (with shadows) added. 

I confirmed in renderdoc that the PR is working as expected (using one long render pass with dynamic viewport and scissor instead of multiple render passes. It just takes an extra 2 ms for rendering shadows (which is confirmed by printing the GPU profile while running)

darksylinc committed:
Always use clears instead of load for directional lights on mobile

Added SUPPORTS_NATIVE_RENDER_AREA_CLEAR to determine whether this is the
best path.
Also removed unused arguments from _render_shadow_pass &
_render_shadow_append.

Must set render_pass_opts_enabled = false to test this new code.

darksylinc commented:
# Updates

We talked with ARM. Since I don't know how much is covered by NDA, I'll just stick to major points that were known beforehand:

1. My guesses were right. The problem is that more Render Passes result in better scheduling overlap of VS and PS work between those passes. The benefit of using better Clear semantics instead of Clear-then-Load is not as big as expected due to certain clever HW optimizations that reduce the expected bandwidth gap between our old and new path.
2. Since there seems to be no difference in Adreno either, this optimization done by TheForge don't seem to have an application in Android and should not be merged.
3. There _might_ still be an application on iOS and M1/M2/etc GPUs. However I couldn't test it since the newest iPhone I have is the iPhone 8, which could not run the demo, seemingly due to HW feature limitations.
4. I went back to the drawing board + some suggestions for ARM, found there was still _something_ we could do, which I will elaborate.


## New optimization

We know that in Vulkan the following is the ideal path for a TBDR (Desktop is another story):

1. Each directional light pass uses Clear semantics instead of Load semantics.
2. Each directional light pass lives in its own Render Pass.

Based on this, I **pushed a new commit** to make the following changes:

1. Added `SUPPORTS_NATIVE_RENDER_AREA_CLEAR` feature flag. When present, it means the optimization is much more likely to be an optimization (and not just a performance net negative). When absent, the old code is used.
   - This is because Metal only allows to clear the whole texture and not just the render area / subregion. Which means MoltenVK (and our Metal backend) must turn to emulation when the clear is only applied to the sub region.
   - Desktop GPUs are more troublesome because some of them (e.g. Vega/RDNA + NVIDIA post Turing or Ampere?) do benefit slightly from Clear semantics, but on the other hand they benefit a lot from the full-region clear operation being done in parallel with async compute. Which is why Forward+ backend still uses the old code.
3. When present, each directional light pass clears its render area.
4. Additionally, removed unused variables (from both Mobile and Forward+ backend).

I know that PRs are encouraged to submit a single commit per PR. I will squash the commits once I get confirmation these changes are ok.

**How to test:** Set `render_pass_opts_enabled` to false and compare before this PR. Performance should be equal or higher, but not lower.

stuartcarnie commented:
Is there anything I could test / analyse for Apple GPUs?

darksylinc commented:
> Is there anything I could test / analyse for Apple GPUs?

Yes!

1. Download [this demo](https://drive.google.com/file/d/1wg1PLTlc6pZs-HzFzQEGAkN3Oo7sqAfB/view?usp=sharing).
2. Compile this PR for iOS (macOS for ARM should work too I suppose). Test running the demo in the following scenarios:
    1. After PR, with `render_pass_opts_enabled = true`.
    2. After PR, with `render_pass_opts_enabled = false`.
    3. Before PR

If we add the native Metal backend (besides MoltenVK), that'd be 6 tests.

Compare the framerate (the demo already has an FPS counter).

clayjohn commented:
I've tested with three devices Pixel 8 Pro, Pixel 7, and Pixel 4. 

I tested all devices with `render_pass_opts_enabled = true` and without this PR. I only tested the Pixel 8 Pro with  `render_pass_opts_enabled = false`, 

**Pixel 8 Pro**
No significant change in any case. 

**Pixel 7**
0.4 mspf improvement with `render_pass_opts_enabled = true`

**Pixel 4**
0.2-0.3 mspf improvement with `render_pass_opts_enabled = true`. To test this I had to set the 3d scale down to 0.4 and turn off all lights except for the directional light.

clayjohn commented on servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:
```diff
@@ -689,6 +689,9 @@ void RenderForwardMobile::_pre_opaque_render(RenderDataRD *p_render_data) {
 	p_render_data->shadows.clear();
 	p_render_data->directional_shadows.clear();
 
+	const bool clear_directional_lights_by_region =
+			RD::get_singleton()->has_feature(RD::SUPPORTS_NATIVE_RENDER_AREA_CLEAR);
```
```suggestion
	const bool clear_directional_lights_by_region = RD::get_singleton()->has_feature(RD::SUPPORTS_NATIVE_RENDER_AREA_CLEAR);
```

I like a short line as much as the next guy, but this is a little excessive

clayjohn reviewed (commented):
The code looks good to me and the results are now consistently the same or better on my test devices. 

I would go ahead and squash the commits and remove the `render_pass_opts_enabled` setting once @stuartcarnie has a chance to test on iOS

stuartcarnie commented:
Heads up that I am on a plane from Phoenix, Arizona to Australia, so I'll be mostly offline for the next 30 hours, in case that is important.

darksylinc commented:
> I've tested with three devices Pixel 8 Pro, Pixel 7, and Pixel 4.
> 
> I tested all devices with `render_pass_opts_enabled = true` and without this PR. I only tested the Pixel 8 Pro with `render_pass_opts_enabled = false`,

Wait, what?

To be clear: I did **NOT** touch the case of `render_pass_opts_enabled = true`.
So if previously it was giving lower perf, it shouldn't be giving better perf now!

The case I was most interested was `render_pass_opts_enabled = false` after vs before PR.

> **Pixel 8 Pro** No significant change in any case.

WHAAAA?

You already tested this phone and toggling `render_pass_opts_enabled` was previously giving you 40 FPS -> 36 FPS.

Did you set the atlas to 4096 instead of the default 2048? (to stress test the shadows) like you did last time?

> **Pixel 7** 0.4 mspf improvement with `render_pass_opts_enabled = true`

Pixel 7 is Mali-G710 MP7 (Valhall). This goes against your findings in Pixel 8 Pro and my findings in Samsung A54 (Mali-G68 MP5, also Valhall).

:confused: 

> **Pixel 4** 0.2-0.3 mspf improvement with `render_pass_opts_enabled = true`. To test this I had to set the 3d scale down to 0.4 and turn off all lights except for the directional light.

Interesting test. I also have an Adreno 640 for testing. I did not see a difference (all tests 6 FPS) but I suspected I was bottlenecked by something else.

I could try the same trick you're doing to see if I see an improvement.

> Heads up that I am on a plane from Phoenix, Arizona to Australia, so I'll be mostly offline for the next 30 hours, in case that is important.

No rush. This PR has been in the pipeline for a month or more.

clayjohn commented:
It turns out I built release templates, but run debug templates. Thanks to our Android buildsystem caching, I managed to run the same benchmark three times. The reason I got a performance differential was likely due to thermal throttling, as I run the scene multiple times and recorded the average frame time from the worst run. Plus I always tested this PR after testing master. 

I ran the benchmark again for each device and got the following consistent results. This time I let the devices cool down and ran the benchmark in different orders.

It seems on Mali devices there is a consistent loss in performance with this PR. While with Adreno there is a slight improvement, but only when `render_pass_opts_enabled = true`.

Pixel 8 Pro
Master: 23.6 mspf
render_pass_opts_enabled = false: 24.2 mspf
render_pass_opts_enabled = true: 25.6 mspf

Pixel 7
Reporting shadow time only since there is too much fluctuation in full frame time
Master: 7.20 mspf
render_pass_opts_enabled = false: 7.78 mspf
render_pass_opts_enabled = true: 9.15 mspf

Pixel 4
Reporting shadow time only since there is too much fluctuation in full frame time
Master: 11.56 mspf
render_pass_opts_enabled = false: 18.70 mspf
render_pass_opts_enabled = true: 11.35 mspf

darksylinc commented:
OK, hard data is hard:

| Device       | Demo                                | Test                                              | MSPF   | FPS   | MSPF (99th %) | FPS (99th %) |
|--------------|-------------------------------------|---------------------------------------------------|--------|-------|---------------|--------------|
| Adreno 640   | Original TPS                        | render_pass_opts_enabled = true                   | 179,15 | 5,58  | 181,06        | 5,52         |
| Adreno 640   | Original TPS                        | render_pass_opts_enabled = false; clear subregion | 181,61 | 5,51  | 183,65        | 5,45         |
| Adreno 640   | Original TPS                        | master branch                                     | 179,06 | 5,58  | 180,83        | 5,53         |
| Adreno 640   | Only directional light + 0,25 scale | render_pass_opts_enabled = true                   | 46,08  | 21,70 | 62,10         | 16,10        |
| Adreno 640   | Only directional light + 0,25 scale | render_pass_opts_enabled = false; clear subregion | 48,52  | 20,61 | 65,10         | 15,36        |
| Adreno 640   | Only directional light + 0,25 scale | master branch                                     | 46,07  | 21,71 | 62,32         | 16,05        |
| Mali-G68 MP5 | Original TPS                        | render_pass_opts_enabled = true                   | 78,62  | 12,72 | 80,04         | 12,49        |
| Mali-G68 MP5 | Original TPS                        | render_pass_opts_enabled = false; clear subregion | 73,19  | 13,66 | 75,02         | 13,33        |
| Mali-G68 MP5 | Original TPS                        | master branch                                     | 73,90  | 13,53 | 76,28         | 13,11        |
| Mali-G68 MP5 | Only directional light + 0,5 scale  | render_pass_opts_enabled = true                   | 42,88  | 23,32 | 49,95         | 20,02        |
| Mali-G68 MP5 | Only directional light + 0,5 scale  | render_pass_opts_enabled = false; clear subregion | 37,95  | 26,35 | 42,20         | 23,70        |
| Mali-G68 MP5 | Only directional light + 0,5 scale  | master branch                                     | 37,48  | 26,68 | 42,36         | 23,61        |

This data was gathered using debugger's profiler:

![image](https://github.com/user-attachments/assets/cb0e81a7-f3f1-4ede-a796-8643643e0c1a)

It conveniently supports exporting to CSV (where I can calculate avg and 99th percentiles). It also provides the most accurate counter for these measurements: "Frame Time".

Well... the data is disheartening: On Adreno 640, TheForge's optimization is indistinguishable from noise. While my optimization causes a slowdown.
On Mali-G68, TheForge's optimization causes a very visible slowdown. While my optimization is this time the one indistinguishable from noise.

We could try to contact ARM for further help, but at this point I'm more inclined to just close the pull request and abort the merge.

clayjohn commented:
> We could try to contact ARM for further help, but at this point I'm more inclined to just close the pull request and abort the merge.

Unfortunately I have to agree. I think we have pushed this PR as far as it can go. It seems like the current code is not always the most efficient, but it is never the least efficient. Anything we change makes the code more complex and risks hitting a performance cliff on one driver or another.
